Memory Caps on Forensics

biscuitWizard opened the pull request:
Nerfing forensics to hold a maximum of 64 prints, dna, residues, or traces (per each) for their respective atoms.

    This should not affect gameplay terribly but *does* provide a strong ceiling to the amount of memory that can be consumed by forensics instead of the previously unlimited amount.

This shouldn't require a changelog.

PsiOmegaDelta commented on code/modules/detectivework/forensics.dm:
```diff
@@ -1,3 +1,5 @@
+#define MAX_EVIDENCE_SIZE 64
```
Should undef defines at the end of file

PsiOmegaDelta commented on code/modules/detectivework/forensics.dm:
```diff
@@ -165,13 +177,20 @@ atom/proc/add_fibers(mob/living/carbon/human/M)
 		LAZYDISTINCTADD(trace_DNA, M.dna.unique_enzymes)
+		// Garbage reduction code.
+		if(trace_DNA.len + 1 > MAX_EVIDENCE_SIZE)
+			trace_DNA.Cut(1, 1 + (trace_DNA.len - MAX_EVIDENCE_SIZE))
```
All these blocks are rather copy-pasty

biscuitWizard replied:
Agreed. But I couldn't figure out a generic list implementation without rewriting forensics & adding a whole new datastructure, and that seemed to outweigh the potential benefit.

comma replied:
Is there really benefit in first place? I really doubt there's an item which has DNA/prints from 64 different people.

PsiOmegaDelta replied:
A macro would probably be pretty easily to implement

biscuitWizard replied:
Could you make a code suggestion of what the macro might look like for this? I'd be happy to incorporate it, but I need to know the pattern.

biscuitWizard replied:
I got it, thanks for the suggestion.

biscuitWizard commented:
Agreed. But I couldn't figure out a generic list implementation without rewriting forensics & adding a whole new datastructure, and that seemed to outweigh the potential benefit.

afterthought2 commented:
Seems like this affects hidden prints, which are an admin investigation aid. Are we really seeing OOM due to this?

PsiOmegaDelta commented:
I believe this is a downstream thing of a persistent "forever" running fork, where such matters is more of a concern.

biscuitWizard commented:
That's correct, this is more a downstream problem, but it shouldn't really have an impact on Bay. It would if we lowered the max size to 16 or 32, but I wanted to do minimal impact to gameplay.

Commit:
Nerfing forensics to hold a maximum of 64 prints, dna, residues, or traces (per each) for their respective atoms.

    This should not affect gameplay terribly but *does* provide a strong ceiling to the amount of memory that can be consumed by forensics instead of the previously unlimited amount

biscuitWizard commented:
Adding as a general note that these lists are on /atom level. Speaking from a best practices position, the means every item has three lists that can infinitely expand at the moment. I don't want to negatively impact gameplay, but that's a LOT of lists that are being added to during every interaction in Bay.

And the larger those lists are, the more time that will be spent adding to them. You could probably limit max evidence size to 16 and not lose any meaningful gameplay, but I leave that decision to you guys.

afterthought2 commented:
Make max evidence size a config variable with a default value of infinity and I'm fine with this. I don't see a need for it on bay live, but if it helps support downstreams then sure, and if at some point it becomes a problem then we can tweak the value.

If I understand correctly, this is a cap on unique fingerprints, basically meaning 64 distinct mobs would have done a relevant interaction. That seems unlikely to begin with, but if it does happen I think we'd rather keep the extra info for admin purposes.

afterthought2 commented:
Closing due to inactivity.
